### ISSUE TITLE
Absolute line numbers

### ISSUE THREAD

**Issue by felixdollack** [jupyterlab/jupyterlab#6718]
I like how Jupiter lab has a slightly better graphical interface than the old notebooks 👍 
I do almost all my daily coding with Juypiter lab now :)

Is there a way to show line numbers by default and make them absolute?
It is really hard to tell someone else less experienced in coding where to look when every cell updates its' execution number and starts at line 1.

Are there plans to support this or are more people in favour for this UI enhancement?

**Comment by felixdollack** [comment]
@meeseeksdev tag "Design and UX"

**Comment by BoPeng** [comment]
Would it be good enough to say line 15.2 where 15 is the index of the cell that be easily located through, for example, the TOC panel? The problem with true global line number is that it is costly to update all cells with the addition or removal of a line.

**Comment by felixdollack** [comment]
@BoPeng yes, a numeration of the panels would help already.
I didn't have tie yet to look at the code of jupyter, but about how costly are we talking?

**Comment by BoPeng** [comment]
The line number is done by codemirror, which does allow the set of a "starting line number" so technically speaking global line number can be done. Triggering an update of the starting line number of all cells with every addition or removal of line during the edition of a cell can be costly (and might not be doable). It is more realistic (and should be good enough) to update starting line numbers with each cell operation (e.g. add/removal of cell, stop editing a cell), just like how the TOC is updated.

**Comment by BoPeng** [comment]
BTW, I am in general against any additional "option" and would make global line number the only choice if it can be done nicely.

**Comment by felixdollack** [comment]
I agree, an update once a cell is executed should be sufficient. I also totally agree that this shouldn't be an option but default and only choice. In my opinion the option should be to hide the line numbers if deemed necessary by the user or cell-type / add-on packeting / etc.

If you can point me towards a starting point I will be happy to make a branch and try if I can make it done nicely ;)

**Comment by BoPeng** [comment]
I do not have a clear idea how to implement this now. I would look in https://github.com/jupyterlab/jupyterlab/tree/master/packages/notebook/src for events emitted after a cell operation. https://codemirror.net/doc/manual.html has doc for `firstLineNumber`.

**Comment by williamstein** [comment]
@jasongrout told me about this thread and suggest I mention what I implemented in CoCalc related to this.  Our motivation for global line information is that CoCalc has multi-user realtime collaboration and integrated (video) chat, so it is absolutely critical for users to be able to refer to locations in a document.

- **Sage worksheet** -- this is similar to Jupyter notebooks, except everything is in one single big CodeMirror editor, and output is in HTML "CodeMirror widgets".   For this, I implemented both global line numbers **and** local per cell line numbers by using two separate CodeMirror gutters, and thinking carefully (so it is very efficient).  Here's what it looks like (and you can try this by making a https://cocalc.com account, then +New --> Sage worksheet):

![image](https://user-images.githubusercontent.com/1276278/63132642-bbd73480-bf76-11e9-85de-539f684431d4.png)

- **CoCalc Jupyter Notebook** - this is our complete reimplementation from scratch of Jupyter that I was driven to create in order to have robust realtime sync.   Each cell has individual line numbers and there is a global cell index (starting from 1) in the **upper right**, next to the timing infomation.  See:

![image](https://user-images.githubusercontent.com/1276278/63133100-77e52f00-bf78-11e9-9a9c-bb2bcc26d116.png)
